### ISSUE TITLE
test failures due to package lock on :common-lisp with SBCL 1.3.13

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#187]
Imported from SourceForge on 2024-07-02 08:18:08
Created by **[tomio-arisaka](https://sourceforge.net/u/tomio-arisaka/)** on 2017-01-08 14:22:35
Original: https://sourceforge.net/p/maxima/bugs/3271

---

run_testsuite failes as follows:
~~~
(%i1) display2d:false$

(%i2) build_info();

(%o2) %build_info("5.39post","2017-01-07 03:05:23",
                  "x86_64-apple-darwin10.8.0","SBCL","1.3.13")
(%i3) run_testsuite();
      . . .
      . . .
Running tests in rtest8: 
********************** Problem 63 (line 220) ***************
Input:
ev(a . % - lambda %, exp )
                        1


Result:
error-catch

This differed from the expected result:
[ x2 - 2 x1 ]
[           ]
[ 2 x1 - x2 ]

********************** Problem 64 (line 222) ***************
Input:
exp : %     = 0
       1, 1


Result:
error-catch     = 0
           1, 1

This differed from the expected result:
x2 - 2 x1 = 0

********************** Problem 66 (line 226) ***************
Input:
solve([exp, %], [x1, x2])


Result:
[]

This differed from the expected result:
         - 1           - 2              1             2
[[x1 = -------, x2 = -------], [x1 = -------, x2 = -------]]
       sqrt(5)       sqrt(5)         sqrt(5)       sqrt(5)

168/171 tests passed

The following 3 problems failed: (63 64 66)
Running tests in rtest12: 79/79 tests passed (not counting 2 expected errors)
Running tests in rexamples: 
********************** Problem 16 (line 62) ***************
Input:
(kill(all), example(charpoly))


Result:
(%i1) a:matrix([3,1],[2,4])
                                   [ 3  1 ]
(%o1)                              [      ]
                                   [ 2  4 ]
(%i2) expand(charpoly(a,lambda))
                                  2
(%o2)                       lambda  - 7 lambda + 10
(%i3) (programmode:true,solve(%))
(%o3)                      [lambda = 5, lambda = 2]
(%i4) matrix([x1],[x2])
                                    [ x1 ]
(%o4)                               [    ]
                                    [ x2 ]
(%i5) ev(a . %-lambda*%,%th(2)[1])
error-catch

This differed from the expected result:
done

136/137 tests passed

The following 1 problem failed: (16)
~~~

Because a new feature of SBCL-1.3.13 is that 
SET triggers package locks on undefined variables.

The example is as follows:
~~~
(%i1) display2d:false$

(%i2) build_info();

(%o2) %build_info("5.39post","2017-01-07 03:05:23",
                  "x86_64-apple-darwin10.8.0","SBCL","1.3.13")
(%i3) lambda : 321;

Maxima encountered a Lisp error:

 Lock on package COMMON-LISP violated when setting the value of LAMBDA while in
 package MAXIMA.
See also:
  The SBCL Manual, Node "Package Locks"
  The ANSI Standard, Section 11.1.2.1.2

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.
(%i4) :lisp #$'(lambda : 321)$

((MSETQ SIMP) LAMBDA 321)
(%i4) :lisp (describe '$lambda)

MAXIMA::$LAMBDA
  [symbol]

Symbol-plist:
  TEXWORD -> "\\lambda"
  SUBC -> NIL
  KIND -> (C)
  ALIAS -> LAMBDA
(%i4) :lisp (describe 'lambda)

COMMON-LISP:LAMBDA
  [symbol]

LAMBDA names a macro:
  Lambda-list: (ARGS &BODY BODY)
  Source file: SYS:SRC;CODE;DEFBOOT.LISP

Symbol-plist:
  TEXWORD -> "\\lambda"
  TR-LISP->LISP -> #<FUNCTION (LAMBDA (FORM)..
  TRANSLATE -> #<FUNCTION (LAMBDA (FORM)..
  FREE-LISP-VARS -> #<FUNCTION (LAMBDA (FORM)..
  REVERSEALIAS -> $LAMBDA
  OPERATORS -> SIMPLAMBDA
  LISP-NO-SIMP -> T
  MFEXPR* -> CONSTLAM
  SB-WALKER::WALKER-TEMPLATE -> SB-WALKER::WALK-LAMBDA
(%i4) :lisp (setf (symbol-value 'lambda) 321)

Maxima encountered a Lisp error:

 Lock on package COMMON-LISP violated when setting the value of LAMBDA while in
 package MAXIMA.
See also:
  The SBCL Manual, Node "Package Locks"
  The ANSI Standard, Section 11.1.2.1.2

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.
(%i4) :lisp *package*

#<PACKAGE "MAXIMA">
(%i4) :lisp (in-package "COMMON-LISP")

#<PACKAGE "COMMON-LISP">
(%i4) :lisp (setf (symbol-value 'lambda) 321)

321
(%i4) :lisp (in-package "MAXIMA")

#<PACKAGE "MAXIMA">
(%i4) :lisp (describe 'lambda)

COMMON-LISP:LAMBDA
  [symbol]

LAMBDA names an undefined variable:
  Value: 321

LAMBDA names a macro:
  Lambda-list: (ARGS &BODY BODY)
  Source file: SYS:SRC;CODE;DEFBOOT.LISP

Symbol-plist:
  TEXWORD -> "\\lambda"
  TR-LISP->LISP -> #<FUNCTION (LAMBDA (FORM)..
  TRANSLATE -> #<FUNCTION (LAMBDA (FORM)..
  FREE-LISP-VARS -> #<FUNCTION (LAMBDA (FORM)..
  REVERSEALIAS -> $LAMBDA
  OPERATORS -> SIMPLAMBDA
  LISP-NO-SIMP -> T
  MFEXPR* -> CONSTLAM
  SB-WALKER::WALKER-TEMPLATE -> SB-WALKER::WALK-LAMBDA
(%i4) lambda;

(%o4) 321
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-02 08:18:10
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-01-08 23:41:28
Original: https://sourceforge.net/p/maxima/bugs/3271/#91a0

---

- **labels**: SBCL --> SBCL, lambda
- **summary**: run_testsuite failes because of SBCL-1.3.13 --> test failures due to package lock on :common-lisp with SBCL 1.3.13

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-02 08:18:12
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-01-08 23:41:29
Original: https://sourceforge.net/p/maxima/bugs/3271/#b57b

---

Verified with SBCL 1.3.13 on Ubuntu 14.04 (x86). The problem is that `lambda` in Maxima parses to LAMBDA in Lisp, not $LAMBDA, so assigning to `lambda` assigns to LAMBDA, and if the package containing LAMBDA is locked, then the error is triggered.

SBCL 1.3.11 doesn't show that behavior, and other varieties of Lisp don't show that error. These variations in behavior are OK according to CLHS 11.1.2.1.2.

I see $LAMBDA has a property ALIAS which has the value LAMBDA (and LAMBDA has the corresponding REVERSEALIAS with value $LAMBDA). 

I see two ways to try to resolve this problem. (1) Change SBCL flags so that assignment to symbols in the package :common-lisp is OK. (2) Remove the alias so that Maxima `lambda` parses to $LAMBDA which is not a :common-lisp symbol.

At this point I don't know whether (1) or (2) is a better way to go.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-02 08:18:14
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-01-09 05:19:40
Original: https://sourceforge.net/p/maxima/bugs/3271/#ed7a

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-02 08:18:16
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-01-09 05:19:40
Original: https://sourceforge.net/p/maxima/bugs/3271/#ab4d

---

Fixed by commit eaacba6. I put in code to unlock :common-lisp if compiling with SBCL. Closing this report as fixed.

I tried two other things: remove the $LAMBDA/LAMBDA alias, and shadow the symbol LAMBDA. In both cases there were errors, so I gave up on that.
